### PR TITLE
Sidebar: Add expand and collapse actions

### PR DESCRIPTION
**Pull request by creativecoder**
#### Changes proposed in this Pull Request

* Refactors my-sites sidebar state to use a keyed reducer for sidebar sections
* Adds expand and collapse actions for sidebar sections

#### Testing instructions

* Load My Sites section of Calypso
* Expand and collapse each section of the sidebar (Site, Design, Tools, Manage)
* With at least 1 sidebar section expanded, reload the page--the section should be expanded once the page has reloaded (if state is hydrated on reload)

Supports #33519

### CODE REVIEW

File: client/my-sites/sidebar/sidebar.jsx [hunk 1]
```diff
@@ -611,8 +607,8 @@ export class MySitesSidebar extends Component {
 				</SidebarMenu>
 
 				<ExpandableSidebarMenu
-					onClick={ this.props.toggleMySitesSidebarSiteMenu }
-					expanded={ this.props.isSiteOpen }
+					onClick={ () => this.props.toggleSection( SIDEBAR_SECTION_SITE ) }
```
Comment (creativecoder) [inline client/my-sites/sidebar/sidebar.jsx]: Is defining an anonymous function within a prop value still considered a performance issue b/c of re-rendering? I noticed the recommendation to not do this has been removed from the [React.js docs on event handling](https://reactjs.org/docs/handling-events.html).

Comment (flootr) [reply client/my-sites/sidebar/sidebar.jsx]: Imho you're good - there's been lots of discussions around this and it highly depends on what you're doing. [Here's a link to what React docs say about that](https://reactjs.org/docs/faq-functions.html#is-it-ok-to-use-arrow-functions-in-render-methods). Also, here's an [in-depth article](https://cdb.reacttraining.com/react-inline-functions-and-performance-bdff784f5578) by _@ryanflorence about this.

@sgomes and @jsnajdr may be able to provide more calypso-focused feedback.

Comment (sgomes) [reply client/my-sites/sidebar/sidebar.jsx]: I'm not aware of anything specific about inline functions. As far as I know, the only potential problem with this sort of code is that you're creating a new instance of the function on every render, and thus causing a re-render of the subcomponent.

But that's not necessarily a problem in itself. Unless it's a super obvious performance issue (like writing an algorithm with exponential complexity), the general advice is to write your code the readable way, and only go back to improve performance if you measure an actual performance problem.

And if that's the case, feel free to ping me and I'd be happy to help you out 🙂 There are multiple ways of getting around this sort of issue (like memoization and `shouldComponentUpdate`) if it comes to that.

Comment (jsnajdr) [reply client/my-sites/sidebar/sidebar.jsx]: In the specific case of this sidebar, inline event handlers don't matter much: the Sidebar component is connected to Redux, and therefore rerenders only when the props returned from `mapStateToProps` change. Therefore, if a rerender of `ExpandableSidebarMenu` is triggered, there's already a valid reason to do it (prop change) and the fact that there's one more changed prop (`onClick`) doesn't cause any extra work.

With all that, I think it's still not a bad habit to make the handlers constant:
```jsx
class Sidebar extends Component {
  toggleSiteSection = () => this.props.toggleSection( SIDEBAR_SECTION_SITE );
  toggleDesignSection = () => this.props.toggleSection( SIDEBAR_SECTION_DESIGN );

  render() {
    return (
      <Expandable onClick={ this.toggleSiteSection } />
      <Expandable onClick={ this.toggleDesignSection } />
    );
  }
}
```

Another, more elegant and less verbose approach is to use a function that returns a function and use `lodash.memoize` to remember the result:
```jsx
class Sidebar extends Component {
  toggleSection = memoize( id => () => this.props.toggleSection( id ) );

  render() {
    return (
      <Expandable onClick={ this.toggleSection( SIDEBAR_SECTION_SITE ) } />
      <Expandable onClick={ this.toggleSection( SIDEBAR_SECTION_DESIGN ) } />
    );
  }
}
```

But as I said, even the inline handlers are quite fine here.

Comment (creativecoder) [reply client/my-sites/sidebar/sidebar.jsx]: ❤️ it, thanks for the knowledge share! I've updated this PR to use `memoize`, which is lightweight enough that to me it seems like a good habit for event handlers that need arguments.

**Comment by Stojdza** [comment]
Hey @creativecoder 👋 This PR caused [some of the e2e tests](https://circleci.com/gh/Automattic/wp-calypso/329542) to fail. This is because tests are relying on `aria-expanded=" false"` when all items in the sidebar are collapsed. I can see it's missing now. Was this intentional? 

Screen Shot 2019-06-16 at 16 11 12

I'll take a look and see how I can fix it so tests can pass again. Please, let me know was this intentional so I can add some permanent solution on e2e side :)

**Comment by creativecoder** [comment]
@Stojdza Removing the aria-expanded attribute was _not_ intentional.

**Comment by creativecoder** [comment]
Also, I can't seem to replicate the bug... I'm still seeing the `aria-expanded` attribute on the `h2` expandable sidebar headings on `master`, and it correctly indicates the state (expanded or not).

**Comment by Stojdza** [comment]
@creativecoder I'm looking into WP.com while I'm proxied. After clicking on My sites, Sidebar is shown and all items are collapsed. For, let's say, Tools, I don't see `aria-expanded="false"`. When I expand it, I see `aria-expanded="true"`. Screenshots:


Screen Shot 2019-06-16 at 21 24 43

Screen Shot 2019-06-16 at 21 24 28

**Comment by creativecoder** [comment]
Thanks--I see the issue now. I'm a bit puzzled about what exactly changed to cause that, but here's an attempted fix: https://github.com/Automattic/wp-calypso/pull/34021.

**Pull request merged**